src/Models/HydrostaticFreeSurfaceModels/mg_implicit_free_surface_solver.jl

```diff
@@ -0,0 +1,205 @@
+using Oceananigans.Solvers: constructors, arch_sparse_matrix, ensure_diagonal_elements_are_present!, update_diag!, unpack_constructors
+using SparseArrays: _insert!
+
+import Oceananigans.Solvers: solve!, precondition!
+import Oceananigans.Architectures: architecture
+
+"""
+    struct MGImplicitFreeSurfaceSolver{V, S, R}
+
+The multigrid implicit free-surface solver.
+
+$(TYPEDFIELDS)
+"""
+mutable struct MGImplicitFreeSurfaceSolver{S, V, F, R, C, D}
+    "The multigrid solver"
+    multigrid_solver :: S
+    "The vertically-integrated lateral areas"
+    vertically_integrated_lateral_areas :: V
+    "The previous time step"
+    previous_Δt :: F
+    "The right hand side of the free surface evolution equation"
+    right_hand_side :: R
+    "The matrix constructors of the linear operator without the `Az / (g Δt²)` term"
+    matrix_constructors :: C
+    "The `Az / g` term"
+    diagonal :: D
+end
+
+architecture(solver::MGImplicitFreeSurfaceSolver) =
+    architecture(solver.multigrid_solver)
+
+"""
+    MGImplicitFreeSurfaceSolver(grid, settings)
+
+Return a solver based on a multigrid method for the elliptic equation
+    
+```math
+[∇ ⋅ H ∇ - 1 / (g Δt²)] ηⁿ⁺¹ = (∇ʰ ⋅ Q★ - ηⁿ / Δt) / (g Δt)
+```
+
+representing an implicit time discretization of the linear free surface evolution equation
+for a fluid with variable depth `H`, horizontal areas `Az`, barotropic volume flux `Q★`, time
+step `Δt`, gravitational acceleration `g`, and free surface at the `n`-th time-step `ηⁿ`.
+"""
+function MGImplicitFreeSurfaceSolver(grid::AbstractGrid, 
+                                    settings, 
+                                    gravitational_acceleration=nothing, 
+                                    placeholder_timestep = -1.0)
+    arch = architecture(grid)
+
+    # Initialize vertically integrated lateral face areas
+    ∫ᶻ_Axᶠᶜᶜ = Field{Face, Center, Nothing}(with_halo((3, 3, 1), grid))
+    ∫ᶻ_Ayᶜᶠᶜ = Field{Center, Face, Nothing}(with_halo((3, 3, 1), grid))
+
+    vertically_integrated_lateral_areas = (xᶠᶜᶜ = ∫ᶻ_Axᶠᶜᶜ, yᶜᶠᶜ = ∫ᶻ_Ayᶜᶠᶜ)
+
+    compute_vertically_integrated_lateral_areas!(vertically_integrated_lateral_areas)
+    fill_halo_regions!(vertically_integrated_lateral_areas)
+
+    # Set some defaults
+    settings = Dict{Symbol, Any}(settings)
+    settings[:maxiter] = get(settings, :maxiter, grid.Nx * grid.Ny)
+    settings[:reltol] = get(settings, :reltol, min(1e-7, 10 * sqrt(eps(eltype(grid)))))
+
+    right_hand_side = Field{Center, Center, Nothing}(grid)
+
+    # initialize solver with Δt = nothing so that linear matrix is not computed; see `initialize_matrix` methods
+    solver = MultigridSolver(Az_∇h²ᶜᶜᶜ_linear_operation!, ∫ᶻ_Axᶠᶜᶜ, ∫ᶻ_Ayᶜᶠᶜ;
+                             template_field = right_hand_side, settings...)
+    # For updating the diagonal
+    matrix_constructors = constructors(arch, solver.matrix)
+    M = grid.Nx * grid.Ny
+    fill_diag!(matrix_constructors, arch, M, M)
```

Comment:
```suggestion
    Nx, Ny = grid.Nx * grid.Ny
    fill_diag!(matrix_constructors, arch, Nx * Ny, Nx * Ny)
```

Reply:
Do you mean to have a coma in line 81 instead of `*`?

Reply:
```suggestion
    Nx, Ny = grid.Nx, grid.Ny
    fill_diag!(matrix_constructors, arch, Nx*Ny, Nx*Ny)
```

@navidcy I think that's what you meant to suggest?